### PR TITLE
Serializer Registration

### PR DESCRIPTION
**Pull request by scrogson**
This PR changes the way serializers are registered. It removes the need for serializers to be added to the `:oauth2` application config.

Instead, serializers are now registered in code with `OAuth2.Client.put_serializer/3`. This allows wrapper libraries to provide custom serializers without users needing to be concerned with configuring the `:oauth2` application.

/cc @tsubery (#88)

### CODE REVIEW

**Comment by snewcomer** [comment]
@scrogson Recently came across a good use case here: https://github.com/ueberauth/ueberauth/issues/87.  

If you don't have time, I would be happy to pick up the flag on this one.  Otherwise, I can review if you are still interested in merging this PR!

**Comment by scrogson** [comment]
Hey @snewcomer,

I've got a new design that I want to work on that will remove the need for a centralized configuration and rely on the configuration to be on the `OAuth2.Client` struct itself. This will allow for greater flexibility and power.

**Comment by snewcomer** [comment]
This is looking great @scrogson.  Lmk if you need an 👁yet!

**Comment by scrogson** [comment]
@snewcomer @doomspork I'd love some eyes on this! :heart:

**Comment by doomspork** [comment]
@scrogson Happy Holidays homie! I will take a peek this weekend and get back to you.  Hope all is well with your family 😁

File: config/config.exs [hunk 1]
```diff
@@ -1,10 +1,9 @@
 use Mix.Config
 
-config :logger, level: :info
+config :logger, level: :debug
```
Comment (doomspork) [inline config/config.exs]: Wouldn't `:info` be better?  Not a deal breaker, just curious.

File: lib/oauth2/client.ex [hunk 2]
```diff
@@ -202,6 +207,43 @@ defmodule OAuth2.Client do
...
+
+  @doc """
+  Un-register a serialization module for a given mime type.
+
+  ## Example
+
+      iex> client = OAuth2.Client.delete_serializer(%OAuth2.Client{}, "application/json")
+      %OAuth2.Client{}
+      iex> OAuth2.Client.get_serializer(client, "application/json")
+      OAuth2.Serializer.Null
```
Comment (doomspork) [inline lib/oauth2/client.ex]: I'm not sure how I feel about `OAuth2.Serializer.Null`, why would you not return `nil`?

File: lib/oauth2/client.ex [hunk 3]
```diff
@@ -202,6 +207,43 @@ defmodule OAuth2.Client do
...
+  """
+  @spec delete_serializer(t, binary) :: t
+  def delete_serializer(%Client{serializers: serializers} = client, mime) do
+    %Client{client | serializers: Map.delete(serializers, mime)}
+  end
+
+  @doc false
+  @spec get_serializer(t, binary) :: atom
+  def get_serializer(%Client{serializers: serializers}, mime) do
+    Map.get(serializers, mime) || OAuth2.Serializer.Null
```
Comment (doomspork) [inline lib/oauth2/client.ex]: See above comment RE: `OAuth2.Serializer.Null` vs `nil`.

**Comment by snewcomer** [comment]
Looks great!

How does this look (references this PR)? https://github.com/ueberauth/ueberauth_google/pull/58

File: lib/oauth2/client.ex (hunk 3 above)
Comment (yordis) [reply lib/oauth2/client.ex]: Could you use `Map.get/3` here instead?

File: config/config.exs (hunk 1 above)
Comment (scrogson) [reply config/config.exs]: Probably. I think this was something that I did to debug something in test.

File: lib/oauth2/client.ex (hunk 2 above)
Comment (scrogson) [reply lib/oauth2/client.ex]: I think this is a great point. I think this is left over from a previous idea and I didn't take the time to remove it. 👍

File: lib/oauth2/client.ex (hunk 3 above)
Comment (scrogson) [reply lib/oauth2/client.ex]: @yordis yep, very true. I'm probably going to end up removing the `Null` serializer altogether.

Comment (doomspork) [reply lib/oauth2/client.ex]: Thanks for the update @scrogson 😁

**Comment by scrogson** [comment]
> How does this look (references this PR)? [ueberauth/ueberauth_google#58](https://github.com/ueberauth/ueberauth_google/pull/58)

@snewcomer hey, sorry for the delay. That PR looks like it will be perfect once this lands. Thanks for that.

I need to clean up a few things and I'll try to get it merged and a new version published.

**Comment by scrogson** [comment]
@doomspork calling for one last review....

**Comment by adamniedzielski** [comment]
@scrogson was it an intentional decision that this PR changes version requirement for `hackney` from `~> 1.7` to `~> 1.13.0`? This is a more restrictive requirement, because `~> 1.7` allows for example `1.15.1`, but `~> 1.13.0` no.

I was trying to update `oauth2` from `0.9.4` to `1.0.0` and this required a downgrade of `hackney` (along with some transitive dependencies coming from `hackney`). At the same time, there's no discussion in this PR related to changing `hackney` version requirement, so I was thinking that I'd better double-check with you. I can also open a new issue if you prefer 😄

**Pull request merged**